File: 2023/day5/day5p2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <string>

int main() {
    std::ifstream file("test.txt");
    std::string line;
    std::vector<int> seeds;

    // Read the first line and extract integers
    if (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string token;
        while (iss >> token) {
            seeds.push_back(token[0]);
        }
    }

    for (int i = 0; i < seeds.size(); ++i) {
        std::cout << seeds[i] << std::endl;
    }
    // std::cout << seeds << std::endl;

    // for (size_t i = 1; i < seeds.size(); ++i) {
    //     if (i % 2 == 1) {
    //         seeds[i] = seeds[i] + seeds[i - 1];
    //     }
    // }

    // std::vector<int> newSeeds;
    // for (size_t i = 1; i < seeds.size(); i += 2) {
    //     for (int j = seeds[i - 1]; j < seeds[i]; ++j) {
    //         newSeeds.push_back(j);
    //     }
    // }
    // seeds = newSeeds;

    // std::vector<int> res;
    // for (int seed : seeds) {
    //     bool changed = false;
    //     file.clear(); // Clear EOF flag
    //     file.seekg(0); // Go back to the start of the file
    //     for (int j = 0; std::getline(file, line) && j < 3; ++j) { }

    //     while (std::getline(file, line)) {
    //         if (line.empty() || isalpha(line[0])) {
    //             continue;
    //         }
    //         std::istringstream iss(line);
    //         std::vector<int> values;
    //         for (int val; iss >> val;) {
    //             values.push_back(val);
    //         }
    //         int start = values[1];
    //         int end = values[1] + values[2] - 1;
    //         int change = values[0] - values[1];

    //         if (start <= seed && seed <= end) {
    //             seed += change;
    //             changed = true;
    //         }
    //     }
    //     res.push_back(seed);
    // }

    // std::cout << *std::min_element(res.begin(), res.end()) << std::endl;
    // return 0;
}

File: 2023/day5/day5p2.py
lines = []
res = []
with open("input.txt", "r") as file:
    while line := file.readline():
        lines.append(line.strip())

seeds = [int(seed) for seed in lines[0].split()[1:]]
# seeds = [3561206012, 483360452]
seeds = [3900000000, 4100000000]
        #4002148012

# for i in range(len(seeds)):
#     if i % 2 == 1:
#         seeds[i] = seeds[i] + seeds[i-1]
seeds[1] = seeds[1] + seeds[0]

# seeds = [seeds[i:i+2] for i in range(0, len(seeds), 2)]

step = 100
seeds = list(range(int(seeds[0]), int(seeds[1]), step))

# for i in range(len(seeds)):
#     seeds[i] = list(range(seeds[i][0], seeds[i][1], step))

def process(seed):
    changed = False
    for i in range(3, len(lines)):
        if changed and lines[i] != "":
            continue
        else:
            changed = False
        start = 0
        end = 0
        change = 0
        if lines[i] == '' or lines[i][0].isalpha():
            continue
        values = lines[i].split()
        start = int(values[1]) 
        end = int(values[1]) + int(values[2]) - 1
        change = int(values[0]) - int(values[1])
        if start <= int(seed) <= end:
            seed += change
            changed = True
        else:
            continue
    return seed

# print(seeds)
ranges = []
smallest = float('inf')
for i in range(len(seeds)):
    location = process(seeds[i])
    if location < smallest:
        smallest = location
        ranges.append(i)

print(ranges[-1])
print(smallest)


# ranges = []
# for i in range(len(seeds)):
#     smallest = float("inf")
#     for j in range(len(seeds[i])):
#         location = process(seeds[i][j])
#         smallest = min(smallest, location)
#     res.append(smallest)
#     ranges.append(i)

# 


#29102684
#17737743
#17729743